src/libostree/ostree-repo-pull.c

```diff
@@ -3566,6 +3570,680 @@ ostree_repo_pull_with_options (OstreeRepo             *self,
+
+/* Validate the given string is potentially a ref name. */
+static gboolean
+is_valid_ref_name (const gchar *ref_name)
+{
+  return (ref_name != NULL && *ref_name != '\0' && g_str_is_ascii (ref_name));
+}
+
+/* Validate @refs is non-%NULL, non-empty, and contains only valid ref names. */
+static gboolean
+is_valid_ref_array (const gchar * const *refs)
+{
+  gsize i;
+
+  if (refs == NULL || *refs == NULL)
+    return FALSE;
+
+  for (i = 0; refs[i] != NULL; i++)
+    {
+      if (!is_valid_ref_name (refs[i]))
+        return FALSE;
+    }
+
+  return TRUE;
+}
+
+/**
+ * ostree_repo_find_remotes:
+ * @self: an #OstreeRepo
+ * @refs: (array zero-terminated=1): non-empty array of refs to find remotes for
+ * @options: (nullable): A GVariant `a{sv}` with an extensible set of flags
+ * @progress: (nullable): an #OstreeAsyncProgress to update with the operation’s
+ *    progress, or %NULL
+ * @cancellable: (nullable): a #GCancellable, or %NULL
+ * @error: return location for a #GError, or %NULL
+ *
+ * Find reachable remote URIs which claim to provide any of the given named
+ * @refs. This will search for configured remotes (#OstreeRepoFinderConfig),
+ * mounted volumes (#OstreeRepoFinderMount) and (if enabled at compile time)
+ * local network peers (#OstreeRepoFinderAvahi). In order to use a custom
+ * configuration of #OstreeRepoFinder instances, call
+ * ostree_repo_finder_resolve_all_async() on them individually.
+ *
+ * Any remote which is found and which claims to support any of the given @refs
+ * will be returned in the results. It is possible that a remote claims to
+ * support a given ref, but turns out not to — it is not possible to verify this
+ * until ostree_repo_pull_from_remotes() is called.
+ *
+ * The returned results will be sorted with the most useful first — this is
+ * typically the remote which claims to provide the most of @refs, at the lowest
+ * latency.
+ *
+ * Each result contains a list of the subset of @refs it claims to provide. It
+ * is possible for a non-empty list of results to be returned, but for some of
+ * @refs to not be listed in any of the results. Callers must check for this.
+ *
+ * Pass the results to ostree_repo_pull_from_remotes() to pull the given @refs
+ * from those remotes.
+ *
+ * No @options are currently supported.
+ *
+ * Returns: (transfer full) (array zero-terminated=1): a potentially empty array
+ *    of #OstreeRepoFinderResults, followed by a %NULL terminator element; or
+ *    %NULL on error
+ * Since: 2017.6
+ */
+OstreeRepoFinderResult **
+ostree_repo_find_remotes (OstreeRepo           *self,
+                          const gchar * const  *refs,
+                          GVariant             *options,
+                          OstreeAsyncProgress  *progress,
+                          GCancellable         *cancellable,
+                          GError              **error)
+{
+  /* TODO: Support disabling certain repo finders. */
```

Comment:
The thing that has me hesitating a bit about the "try everything" model - it feels like we'll likely end up needing to hardcode things like "prefer USB finder first".  But even there, I worry about needing some mechanism for the client app to configure individual finders.  We can theoretically do any arbitrary extensions later via new API, but it's worth thinking about a bit now maybe?